service/docker.go

```diff
@@ -126,6 +94,101 @@ func (d *Docker) StartWithCancel() (*StartedService, error) {
 		d.removeContainer(ctx, d.Client, container.ID)
 		return nil, fmt.Errorf("no bindings available for %v", selenium)
 	}
+	seleniumHostPort, vncHostPort := d.getHostPort(stat, selenium, vnc)
+	u := &url.URL{Scheme: "http", Host: seleniumHostPort, Path: d.Service.Path}
+	serviceStartTime := time.Now()
+	err = wait(u.String(), d.StartupTimeout)
+	if err != nil {
+		d.removeContainer(ctx, d.Client, container.ID)
+		return nil, fmt.Errorf("wait: %v", err)
+	}
+	log.Printf("[%d] [SERVICE_STARTED] [%s] [%s] [%v]\n", d.RequestId, d.Service.Image, container.ID, time.Since(serviceStartTime))
+	log.Printf("[%d] [PROXY_TO] [%s] [%s] [%s]\n", d.RequestId, d.Service.Image, container.ID, u.String())
+	s := StartedService{
+		Url:         u,
+		ID:          container.ID,
+		VNCHostPort: vncHostPort,
+		Cancel:      func() { d.removeContainer(ctx, d.Client, container.ID) },
+	}
+	return &s, nil
+}
+
+func (d *Docker) getPortConfig() (*portConfig, error) {
+	selenium, err := nat.NewPort("tcp", d.Service.Port)
+	if err != nil {
+		return nil, fmt.Errorf("new selenium port: %v", err)
+	}
+	exposedPorts := map[nat.Port]struct{}{selenium: {}}
+	var vnc nat.Port
+	if d.VNC {
+		vnc, err = nat.NewPort("tcp", "5900")
+		if err != nil {
+			return nil, fmt.Errorf("new vnc port: %v", err)
+		}
+		exposedPorts[vnc] = struct{}{}
+	}
+	portBindings := nat.PortMap{}
+	if d.IP != "" || !d.InDocker {
+		portBindings[selenium] = []nat.PortBinding{{HostIP: "0.0.0.0"}}
+		if d.VNC {
+			portBindings[vnc] = []nat.PortBinding{{HostIP: "0.0.0.0"}}
+		}
+	}
+	return &portConfig{
+		SeleniumPort: selenium,
+		VNCPort:      vnc,
+		PortBindings: portBindings,
+		ExposedPorts: exposedPorts}, nil
+}
+
+func (d *Docker) getTimeZone() *time.Location {
+	timeZone := time.Local
+	if d.TimeZone != "" {
+		tz, err := time.LoadLocation(d.TimeZone)
+		if err != nil {
+			log.Printf("[%d] [BAD_TIMEZONE] [%s]\n", d.RequestId, d.TimeZone)
+		} else {
+			timeZone = tz
+		}
+	}
+	return timeZone
+}
+
+func (d *Docker) getEnv() []string {
+	env := []string{
+		fmt.Sprintf("TZ=%s", d.getTimeZone()),
+		fmt.Sprintf("SCREEN_RESOLUTION=%s", d.ScreenResolution),
+		fmt.Sprintf("ENABLE_VNC=%v", d.VNC),
+	}
+	env = append(env, d.Service.Env...)
+	return env
+}
+
+func (d *Docker) getShmSize() int64 {
+	if d.Service.ShmSize > 0 {
+		return d.Service.ShmSize
+	}
+	return int64(268435456)
+}
+
+func (d *Docker) getContainerHostname() string {
+	if d.ContainerHostname != "" {
+		return d.ContainerHostname
+	}
+	return "localhost"
+}
+
+func (d *Docker) getExtraHosts() []string {
```

Comment:
I think it will be better if values from `browsers.json` are merged with value from `hostEntries` capability. I see use cases when we need "permanent" values for host entries (that are valid for all browser sessions) + values that are changed dynamically and need to be passed via capability.
@vania-pooh

Reply:
I have also started working on a PR for the same feature (just wanted to play with golang a bit). Just in case you need to update `docs/special-capabilities.adoc` (just replace ''' with triple backtick):
```
=== Extra hosts

You can specify additional host entries that are added to /etc/hosts file inside a browser container (if case you need different hosts file entries for each browser session)
'''
hostEntries: my-cool-app:192.168.1.1
'''
Extra hosts are specified as a comma-separated list of `<hostname>:<ip_address>` pairs.
[NOTE]
====
Host entries, passed via this capability will have precedence over the values from <<Browsers Configuration File>> (in case of duplicate entries for the same hostname)
====

```

Reply:
@vgrigoruk we usually add documentation after releasing the feature. But anyway will consider your notes. How would you merge the hosts? E.g. having in `browsers.json`:
```
my-host.example.com:192.168.0.1
```
... passing the following in capability:
```
my-host.example.com:10.0.0.1
```
Should we have two lines in resulting `hosts` or need to compare host names and replace one from `browsers.json`?

Reply:
Actually, both approaches are valid:
1) Compare host-names and if there are duplicate "hostname" found, use the one from capability
2) Prepend hostEntries from capabilities before hostEntires from browsers.json, as the first entry from /etc/hosts always have a priority. So, `my-host.example.com` will be resolved to `10.0.0.1` if we have the following in /etc/hosts:
```
my-host.example.com:10.0.0.1
my-host.example.com:192.168.0.1
```

Reply:
Ok, let's use the second pattern as it is simpler to implement.